crates/rspack_core/src/tree_shaking/optimizer.rs

```diff
@@ -493,12 +493,13 @@ impl<'a> CodeSizeOptimizer<'a> {
                 .and_then(|module| module.as_normal_module())
                 .map(|normal_module| normal_module.source())
               {
-                Some(NormalModuleSource::BuiltFailed(_)) => {
-                  // We know that the build output can't run, so it is alright to generate a wrong tree-shaking result.
-                  continue;
+                Some(NormalModuleSource::Unbuild) => {
+                  panic!("Failed to build module {module_identifier}");
                 }
                 Some(_) => {
-                  panic!("Failed to ast of {dependency_id:?}")
+                  // If module is failed to build, we know that the build output can't run, so it is alright to generate a wrong tree-shaking result.
+                  // Also, if the referenced module of the dependency is not found, then it might failed to factorize in the first place. So let's skip it.
```

Comment:
`Module` might be not available when `dependency` is available if module factorization is failed. For example, empty dependencies are not able to create corresponding modules.